### ISSUE TITLE
How can I add a dh_* tool like dh_dkms to debhelper?

### ISSUE THREAD

**Issue by lucmann** [Debian/debhelper#5]
I am stuck in creating a debian package on my WSL2 CentOS7 distro. It failed with 
```
$ dpkg-buildpackage -b -us -uc
dpkg-buildpackage: info: source package hello-world
dpkg-buildpackage: info: source version 0.1-1
dpkg-buildpackage: info: source distribution unstable
dpkg-buildpackage: info: source changed by unknown <lucmann@qq.com>
dpkg-buildpackage: info: host architecture amd64
 dpkg-source --before-build hello-world-0.1
dpkg-checkbuilddeps: error: Unmet build dependencies: debhelper (>= 8.0.0) dkms
dpkg-buildpackage: warning: build dependencies/conflicts unsatisfied; aborting
dpkg-buildpackage: warning: (Use -d flag to override.)
```
But debhelper and dkms has been installed and the same building has been okay if I do it on my another WSL2 distro Ubuntu 20.04. So I try to check what's difference between these two of environment. 

When I listed all of dh_* commands on the Ubuntu 20.04:
```
$ dh_
dh_auto_build                   dh_icons                        dh_installmenu                  dh_movefiles
dh_auto_clean                   dh_install                      dh_installmime                  dh_perl
dh_auto_configure               dh_installcatalogs              dh_installmodules               dh_perl_openssl
dh_auto_install                 dh_installchangelogs            dh_installpam                   dh_prep
dh_auto_test                    dh_installcron                  dh_installppp                   dh_pypy
dh_autoreconf                   dh_installdeb                   dh_installsystemd               dh_python2
dh_autoreconf_clean             dh_installdebconf               dh_installsystemduser           dh_python3
dh_autotools-dev_restoreconfig  dh_installdirs                  dh_installtex                   dh_shlibdeps
dh_autotools-dev_updateconfig   dh_installdocs                  dh_installtmpfiles              dh_sphinxdoc
dh_bash-completion              dh_installemacsen               dh_installudev                  dh_strip
dh_bugfiles                     dh_installexamples              dh_installwm                    dh_strip_nondeterminism
dh_builddeb                     dh_installgsettings             dh_installxfonts                dh_systemd_enable
dh_clean                        dh_installifupdown              dh_installxmlcatalogs           dh_systemd_start
dh_compress                     dh_installinfo                  dh_link                         dh_testdir
dh_dkms                         dh_installinit                  dh_lintian                      dh_testroot
dh_doxygen                      dh_installinitramfs             dh_listpackages                 dh_ucf
dh_dwz                          dh_installlogcheck              dh_make                         dh_update_autotools_config
dh_fixperms                     dh_installlogrotate             dh_makeshlibs                   dh_usrlocal
dh_gconf                        dh_installman                   dh_md5sums                      dh_virtualenv
dh_gencontrol                   dh_installmanpages              dh_missing
```

while all of dh_* commands on CentOS7:
```
$ dh_
dh_auto_build               dh_icons                    dh_installinfo              dh_installudev              dh_shlibdeps
dh_auto_clean               dh_install                  dh_installinit              dh_installwm                dh_strip
dh_auto_configure           dh_installcatalogs          dh_installinitramfs         dh_installxfonts            dh_systemd_enable
dh_auto_install             dh_installchangelogs        dh_installlogcheck          dh_link                     dh_systemd_start
dh_auto_test                dh_installcron              dh_installlogrotate         dh_lintian                  dh_testdir
dh_bugfiles                 dh_installdeb               dh_installman               dh_listpackages             dh_testroot
dh_builddeb                 dh_installdebconf           dh_installmanpages          dh_make                     dh_ucf
dh_clean                    dh_installdirs              dh_installmenu              dh_makeshlibs               dh_update_autotools_config
dh_compress                 dh_installdocs              dh_installmime              dh_md5sums                  dh_usrlocal
dh_dwz                      dh_installemacsen           dh_installmodules           dh_missing                  dh_virtualenv
dh_fixperms                 dh_installexamples          dh_installpam               dh_movefiles
dh_gconf                    dh_installgsettings         dh_installppp               dh_perl
dh_gencontrol               dh_installifupdown          dh_installsystemd           dh_prep
```
They are fewer than those on Ubuntu 20.04. And dh_dkms is very absent.  

```
$ sudo  yum install debhelper
[sudo] password for luc:
Loaded plugins: fastestmirror
Loading mirror speeds from cached hostfile
 * epel: mirrors.bfsu.edu.cn
Package debhelper-11.4-2.el7.noarch already installed and latest version
```
```
$ sudo apt install debhelper
[sudo] password for luc:
Reading package lists... Done
Building dependency tree
Reading state information... Done
debhelper is already the newest version (12.10ubuntu1).
```
**Should I upgrade my centos7? or is there a way to add dh_dkms to my current debhelper?**

Thanks!

**Comment by nthykier** [comment]
Luc Ma:
> I am stuck in creating a debian package on my WSL2 CentOS7 distro.

Hi,

This is quite an uncommon problem given that most people build debian
packages in a Debian-based distro rather than a redhat/centos based distro.

I would like to know more about your reasons for using a CentOS7 distro
to ensure we are not looking at an [XY Problem], where you are really
trying to solve a different problem and we should instead be looking at
a different approach (like "Use a Debian-based distro like Ubuntu" or
"Look for the CentOS tooling to solve your problem rather than the
Debian based one")

[XY Problem]: https://en.wikipedia.org/wiki/XY_problem

That said, I have taken a bit of time to answer your initial question
all the same.

> It failed with 
> ```
> $ dpkg-buildpackage -b -us -uc
> [...]
>  dpkg-source --before-build hello-world-0.1
> dpkg-checkbuilddeps: error: Unmet build dependencies: debhelper (>= 8.0.0) dkms
> dpkg-buildpackage: warning: build dependencies/conflicts unsatisfied; aborting
> dpkg-buildpackage: warning: (Use -d flag to override.)
> ```

This error from dpkg-checkbuilddeps is presumably because
dpkg-checkbuilddeps assumes it is on a Debian system and is checking the
dpkg "database" of installed packages.  Given you are on a CentOS, then
I presume the dpkg database will be empty as debhelper and dkms is
likely installed via a yum/dnf repo.

Unfortunately, there is not a 1:1 between redhat/rpm based packages and
the Debian packages.  Notably "dkms" has a key difference.

Note that this part technically out of scope for debhelper, so there is
not a lot I can help you with.

> But debhelper and dkms has been installed and the same building has been okay if I do it on my another WSL2 distro Ubuntu 20.04. So I try to check what's difference between these two of environment. 
> 

Ubuntu is based on Debian and will correctly flag the packages as
installed in the place where dpkg-checkbuilddeps can see them.

> When I listed all of dh_* commands on the Ubuntu 20.04:
> ```
> [...]
> ```
> 
> while all of dh_* commands on CentOS7:
> ```
> $ dh_
> [...]
> ```
> They are fewer than those on Ubuntu 20.04. And dh_dkms is very absent.  
> 
> [...]
> ```
> **Should I upgrade my centos7? or is there a way to add dh_dkms to my current debhelper?**
> 
> Thanks!
> 

The dh_dkms tool is provided by the dkms package in Debian and not
debhelper.  This is because it is a "third-party" extension to debhelper
provided by the Debian package maintainer.  Unfortunately, it seems that
the rpm-based version of dkms does not provide dh_dkms - but then dh_*
is a Debian specific package tool.

If you really want to go down this route, you either have to provide
dh_dkms in your PATH manually or convince the CentOS maintainers for
either dkms or debhelper to provide the extra tool.  Though likely they
will find it awkward and strange since neither of them are likely to see
it is a good fit for their package.


To summarize: I think we might be entirely on the wrong track and I
would like to know what problem you are trying to solve. Without knowing
that, I cannot tell if I am leading you in the right direction or
instead directing you on a while goose-chase with no useful result in
the end.

**Comment by lucmann** [comment]
@nthykier thanks, in fact my initial problem is all about building RPM and Debian packages on a single platform as a Jenkins slave node. Because I can't be bothered to start two separate nodes for RPM and Debian packages respectively. Besides it works fine for other normal Debian package except DKMS based source kernel module package on a CentOS. It now appears that it is my best choice to start a qemu thing as a separate node dedicated for creating Debian packages.
